### PR TITLE
First cut of ARMCC support for MESON.

### PR DESCRIPTION
**Pull request by mbedarka**
- Support for compiling .s files with ArmCompiler, added 's' to self.can_compile_suffixes[] of ArmCompiler class
- Ignore .rsp files for ARMCC as these are not valid input files for armcc conpiler
- Removed the dependency file generation related compiler arguments. Made temporary change to to return an empty list from 'get_dependency_gen_args' definition in ArmCompiler class for now.
- Added support to add ability to support Gerrit reviews on meson
- Add support for .axf output file format and include id name in ArmCompiler class __init__ for being accessed in build.py for determining the arm compiler and setting the .axf output file format.
- Ran run_tests.py to make sure no failures were introduced with this commit.

Commit 2 changes
- Fixed SideCI failures
- Updated arm compiler constructor to error if is_cross=False
- Added example cross files to cross directory

Commit 3 changes
- Added Cross Test log .txt
-  For now the crosstests fails for tests related to cpp, shared library(.so) and some custom target 
     tests.  
- Debugging .so and custom target tests are failing. 
- Updated wrap.py to remove changes from initial commit.

Commit 3 changes
- Added support for CPP compiler.

### CODE REVIEW

File: mesonbuild/compilers/compilers.py [hunk 1]
```diff
@@ -112,6 +112,12 @@ def is_library(fname):
                           'debugoptimized': ['-O2', '-g'],
                           'release': ['-O3'],
                           'minsize': ['-Os', '-g']}
+arm_buildtype_args = {'plain': [],
+                      'debug': ['-O0', '-g'],
+                      'debugoptimized': ['-O2', '-g'],
+                      'release': ['-O2'],
+                      'minsize': ['-Os', '-g'],
+                     }
```
Comment (ignatenkobrain) [inline mesonbuild/compilers/compilers.py]: __[Flake8]__

[E124] closing bracket does not match visual indentation

File: mesonbuild/compilers/compilers.py [hunk 2]
```diff
@@ -133,6 +139,12 @@ def is_library(fname):
                                  'release': ['-Wl,-O1'],
                                  'minsize': [],
                                  }
+arm_buildtype_linker_args = {'plain': [],
+                             'debug': [],
+                             'debugoptimized': [],
+                             'release': [],
+                             'minsize': [],
+                            }
```
Comment (ignatenkobrain) [inline mesonbuild/compilers/compilers.py]: __[Flake8]__

[E124] closing bracket does not match visual indentation

File: mesonbuild/wrap/wrap.py [hunk 3]
```diff
@@ -203,15 +203,17 @@ def get_git(self, p):
...
                                           cwd=checkoutdir)
         else:
             subprocess.check_call(['git', 'clone', p.get('url'),
                                    p.get('directory')], cwd=self.subdir_root)
             if revno.lower() != 'head':
-                subprocess.check_call(['git', 'checkout', revno],
-                                      cwd=checkoutdir)
+                if subprocess.call(['git', 'checkout', revno], cwd=checkoutdir) != 0:
+                    subprocess.check_call(['git', 'fetch', p.get('url'), revno], cwd=checkoutdir)
+                    subprocess.check_call(['git', 'checkout', revno],
```
Comment (jpakkane) [inline mesonbuild/wrap/wrap.py]: These changes should not be in this MR as they have nothing to do with the Arm compiler as such. Please move them to their own MR.

**Comment by jpakkane** [comment]
Is armcc only a cross compiler (arm's web site is not really responding ATM)? If yes then the arm compiler should error out in its constructor if it is being invoked with `is_cross=False`.

What fraction of the cross test suite is it possible to run with this? It would be good to get a log of a full run of the cross test suite (we can ignore all failures coming from C++ tests). Please also add a an example cross file in the cross directory so people can copy that easily.

File: mesonbuild/wrap/wrap.py (hunk 3 above)
Comment (mbedarka) [reply mesonbuild/wrap/wrap.py]: Yes I will back these out. I realized later that this should be a separate change. Sorry for the mess up.

File: mesonbuild/compilers/compilers.py (hunk 2 above)
Comment (mbedarka) [reply mesonbuild/compilers/compilers.py]: Agreed. Will submit new patch change to reflect these changes

File: mesonbuild/compilers/compilers.py (hunk 1 above)
Comment (mbedarka) [reply mesonbuild/compilers/compilers.py]: Agreed. Will submit new patch change to reflect these changes

**Comment by mbedarka** [comment]
@jpakkane - How can I upload the Full log of the cross test suite?
[meson-cross-test-run.txt](https://github.com/mesonbuild/meson/files/1791467/meson-cross-test-run.txt)
[meson-cross-test-run.xml.txt](https://github.com/mesonbuild/meson/files/1791468/meson-cross-test-run.xml.txt)
[cross_test_log.txt](https://github.com/mesonbuild/meson/files/1791469/cross_test_log.txt)

Never mind I got it :)

**Comment by fabio-porcedda** [comment]
I've compared with my pull request and i found is missing the C++ support.
Also --buildtype minsize doesn't work because -Os is a gcc option.

I think that the "arm_buildtype_args" are just a copy of gcc parameters and are not suited for armcc.

I think that my "arm_buildtype_args" are better and works!
arm_buildtype_args = {'plain': [],
                      'debug': ['-O0', '--debug'],
                      'debugoptimized': ['-O1', '--debug'],
                      'release': ['-O3', '-Otime'],
                      'minsize': ['-O3', '-Ospace'],
                      }

IMHO my arm_buildtype_args are better because:
- armcc default options are "-O2 -Ospace",
- maximum optimization are "-O3" 
- "Otime" performs optimizations to reduce execution time at the expense of a possible increase in image size.
- -O1 The compiler only performs optimizations that can be described by debug information.  If used with 
   --debug, this option gives a generally satisfactory debug view with good code density.

IMHO i prefer to have a new class for the armlink instead of customizing VisualStudioLinker.

I suggest to take a look at my pull request and taking the parts that are better or the other way around.

**Comment by mohdamerkhalidi** [comment]
Hi @fabio-porcedda ,
  Yes, you are right about the arm_buildtype_args. We need to modify that.
 " **i prefer to have a new class for the armlink instead of customizing VisualStudioLinker**"
   We are not sure we understand this - because we have not modified the VisualStudioLinker.

**Comment by mbedarka** [comment]
@jpakkane - Updated to have support for CPP. 
[cross_test_log.txt](https://github.com/mesonbuild/meson/files/1813932/cross_test_log.txt)
[meson-cross-test-run.txt](https://github.com/mesonbuild/meson/files/1813933/meson-cross-test-run.txt)

**Comment by mbedarka** [comment]
@fabio-porcedda - Please view updated and merged files. Let me know if you spot anything incorrect. We have been able to build with the ARMCC compiler for both c and cpp ( thanks for the cpp additions ). Our compiler does not have support for .so and also does not support building for PCH support.

**Comment by jpakkane** [comment]
Don't add build logs to the MR. Use pastebin or something else instead.

File: cross/armcc.txt [hunk 4]
```diff
@@ -0,0 +1,11 @@
+[binaries]
+c = 'armcc'
+cpp = 'armcc'
+ar = 'armar'
+strip = 'armar'
+
+[host_machine]
+system = 'WINDOWS'
```
Comment (jpakkane) [inline cross/armcc.txt]: These should be in lower case. However this seems to be for a bare metal deployment so it should not say Windows but instead something like `baremetal`.

File: mesonbuild/compilers/compilers.py [hunk 5]
```diff
@@ -113,6 +113,13 @@ def is_library(fname):
                           'release': ['-O3'],
                           'minsize': ['-Os', '-g']}
 
+arm_buildtype_args = {'plain': [],
+                      'debug': ['-O0', '--debug'],
+                      'debugoptimized': ['-O1', '--debug'],
```
Comment (jpakkane) [inline mesonbuild/compilers/compilers.py]: Why is this -O1 instead of -O2? Is ARM somehow nonstandard?

File: mesonbuild/compilers/compilers.py [hunk 6]
```diff
@@ -113,6 +113,13 @@ def is_library(fname):
...
+                      'release': ['-O3', '-Otime'],
+                      'minsize': ['-O3', '-Ospace'],
```
Comment (jpakkane) [inline mesonbuild/compilers/compilers.py]: Similarly should this be only `-Ospace`? Every other compiler I know works so that only the last optimization type flag has any effect.

File: mesonbuild/environment.py [hunk 7]
```diff
@@ -791,7 +799,7 @@ def detect_static_linker(self, compiler):
             if 'lib' in linker or 'lib.exe' in linker:
                 arg = '/?'
             else:
-                arg = '--version'
+                    arg = '--version'
```
Comment (jpakkane) [inline mesonbuild/environment.py]: Stray edit?

File: cross/armcc.txt [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+cpu_family = 'arm'
+cpu = 'Cortex-M0+'
```
Comment (jpakkane) [inline cross/armcc.txt]: Again, the convention is to have these in lower case (in Meson tree, people can set these to anything they want in their own projects).

We have also by convention had `cpu` as the exact processor string that the operating system reports. Something like `armv7hlv`. I'm not familiar enough with this use case to say if this is a good policy to follow in this particular case. You can choose what to have here, just throwing this information out here to help you make the call.

File: mesonbuild/compilers/compilers.py (hunk 5 above)
Comment (mbedarka) [reply mesonbuild/compilers/compilers.py]: In our experience with ARM compiler, -O1 option does restricted optimizations for debugging and O2 will provide high optimization which is not suitable to debug with a debugger. Hence the -O1 level for debug Optimized.

File: mesonbuild/compilers/compilers.py (hunk 6 above)
Comment (mbedarka) [reply mesonbuild/compilers/compilers.py]: Reading through the ARM documentation and help menu there are 2 options 1 being the level of optimizations and 2 being the type of optimization. Hence we have 1 option each for level and type

**Comment by mbedarka** [comment]
@fabio-porcedda - Hi Fabio, please review the changes as requested by you. Just had one question regarding the ARMLinker class you have added. We are using the armlink tool as a linker tool while you are using armcc. Why not just use armlink? Any specific reason you are using armcc for this? 
Also why have you created the ARMLinker class? its unclear to me what the intention is here. I have not merged this section into the commit for now.
Thanks

**Comment by jpakkane** [comment]
The cross test log at the top is very out of date. Can you post an up to date build log and numbers on what percentage of the tests pass?

**Comment by jpakkane** [comment]
Ping. We are approaching dev freeze for the next release and it would be nice to get this in before that.

**Comment by mbedarka** [comment]
@jpakkane - Sorry for the delay in getting the logs uploaded. Here is a set of the latest logs from our latest run.
Highlights are teh highlights of the cross tests
Total passed cross tests: 132
Total failed cross tests: 42
Total skipped cross tests: 5

Assuming that we have a test case size of 179 tests and all failed and skipped test are failures the passing percentage is **74% Passing**

Assuming that the test case size is only the tests that either passed of failed excluding the skipped ones ( total test case size 174), the passing percentage is  **77% Passing**

Hope this is all the information you are looking for. Please let me know if there are any other concerns we should be looking to address. 

Just an FYI we are internally working on a list of issues/enhancements we would like to contribute. Once I have that list compiled I will start to push those changes/updates in as well. 
Thanks

[cross_test_log.txt](https://github.com/mesonbuild/meson/files/1903945/cross_test_log.txt)
[meson-cross-test-run.txt](https://github.com/mesonbuild/meson/files/1903946/meson-cross-test-run.txt)

**Comment by jpakkane** [comment]
Do you want to get this in to 0.46.0 (to which dev freeze was yesterday) or to the next release?

I'm ok with adding this for 0.46.0 so people can use it but we need your input and cooperation for that.

**Comment by mbedarka** [comment]
@jpakkane  - Yes it would be great if we can get this in 0.46.0. We are here to support and help! We are already using this internally for development. Users have already raised a few issues which we will be addressing via patches.
